compiler/rustc_metadata/src/locator.rs

```diff
@@ -1080,7 +1080,7 @@ impl CrateError {
                                 locator.triple
                             ));
                         }
-                        if missing_core && std::env::var("RUSTUP_HOME").is_ok() {
+                        if missing_core && std::env::var("RUST_RECURSION_COUNT").is_ok() {
                             err.help(&format!(
                                 "consider downloading the target with `rustup target add {}`",
                                 locator.triple
```

Comment:
An option would be to use the env var only to be more specific on the `help`:

```suggestion
                        if missing_core {
                            err.help(&format!(
                                "{}consider downloading the target with `rustup target add {}`",
                                if std::env::var("RUST_RECURSION_COUNT").is_ok() {
                                    "if you have installed `rustc` with `rustup`, 
                                } else {
                                    ""
                                },
                                locator.triple
```

An even better check would be to use [`which`](https://crates.io/crates/which) (or similar) to see if `rustup` is available at all before giving the suggestion (but then we could get in the weird situation where `rustc` was installed by, let's say `apt`, but `rustup` is available and the suggestion would be wrong).